Add dropdown display style to Attribute Filter block

Aljullu opened the pull request:
Part of #1134 and #1135.

This PR implements the dropdown display style for the _Filter Products by Attribute_ block with the _Query Type_ set to `OR` (see #1134 for the difference between `OR` and `AND`).

I plan to implement the `AND` format in a follow-up, but wanted to get feedback early before starting with it.

#### Accessibility
- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots
![Peek 2019-11-25 20-03](https://user-images.githubusercontent.com/3616980/69569869-db8ee780-0fbe-11ea-80f9-52fd95c9be20.gif)

### How to test the changes in this Pull Request:

1. Create a post with a _Filter Products by Attribute_ block and select the _Dropdown_ option in _Display Style_.
2. Preview the post and interact with the filter (search terms, add them, remove them, repeat only using the keyboard, using a screen reader, etc.).
3. Verify everything works as expected.

### Follow-ups
* As mentioned, the `AND` dropdown format will be implemented in a follow-up.
* I didn't implement the _Apply_ button that can be seen in the mockups. I would like to discuss with @jwold not showing it or making it optional as in the _Price Filter_, so instant updates can happen.

### Changelog

> Add dropdown display style to Filter Products by Attribute block.

nerrad commented:
I'm getting validation errors on existing usage of Filter by Attribute block.  Since this block has been released, I think we need to ensure there are no validation errors.

Aljullu commented:
You're right @nerrad! That should be fixed in the last commit (fae6b717b4aa114b3db44392eba3c4ad577b9a98).

nerrad commented on assets/js/base/components/checkbox-list/index.js:
```diff
@@ -97,21 +97,23 @@ const CheckboxList = ( {
 			<Fragment>
 				{ options.map( ( option, index ) => (
-					<Fragment key={ option.key }>
+					<Fragment key={ option.value }>
```
I'm curious why this change?  I know you changed the expected shape from `option.key` to `option.value` (so `option.key` no longer exists`) Is this just to make it more consistent (being it is the _value_ for the option)?

Aljullu replied:
There is no reason in fact, I modified it initially because the Downshift example was using `value` instead of `key` and to get started it was easier to modify the data structure of `options`.

Later, I didn't undo the change because both `key` and `value` sound good to me. But I don't mind undoing it if you think `key` was better.

nerrad commented on assets/js/base/components/dropdown-selector/index.js:
```diff
@@ -0,0 +1,154 @@
+	onChange = () => {},
+	options = [],
+} ) => {
+	const inputRef = useRef( null );
```
The same `inputRef` value is implemented on multiple inputs.  Was that intentional?

Aljullu replied:
Good catch! That was probably a copy-&-paste error because `<DropdownSelectorInputWrapper>` was not using the `inputRef` prop. Removed it in 7f77564fe8681ad172ba2a4851d72e4308d9591e.

nerrad commented:
Oh I also forgot to add that when I run `npm install` in this branch, I get a `package-lock.json` diff (which includes the downshift install).  Are you pushing package-lock.json changes?

Aljullu commented:
Thanks for the review @nerrad. I fixed all issues I found in AXE related to this PR and opened an issue to audit the other blocks: #1281.

> The selected item chips do not have a high enough color contrast ratio for accessibility:

Fixed in 5a6b7eb54e188560ed2bb49dc42030c9ba357efe.

> When I use the keyboard to remove a selected item (chip) by tabbing over the chip and hitting enter, the focus is lost from the input (and I have to hit tab again to return focus to the input). However backspacing to remove a chip works as expected.

Fixed in 85c3cfd75a7eb6eb9d1a85cd35a903305d2535a4.

> Oh I also forgot to add that when I run `npm install` in this branch, I get a `package-lock.json` diff (which includes the downshift install).  Are you pushing package-lock.json changes?

I was sure I added it but it looks like I didn't. :man_facepalming:  Done in f313d6520edc70a77fe33f202cfd0109158ddc5b.

Aljullu commented:
> However I still see this:

Good catch! I missed the case when menu was closed. Should be fixed in f5b7e57c4c602a555f741730f8510bee48ae2f68.

> Also, there's still a loss of focus when tabbing and hitting enter to remove a selected item - here's a gif that might help (but it doesn't show the key presses...highlighted chip is when the tab is focused on it).

The GIF was helpful, thank you! I think I misunderstood you in the previous message. This issue should be fixed in de78883306e2a399f771ff31b20522af5459ac25.

jwold commented:
@Aljullu I think having the apply button as optional (but turned off), makes a lot of sense. Great work!
